src/stream/src/executor/backfill/utils.rs

```diff
@@ -79,32 +89,147 @@ impl BackfillState {
     pub(crate) fn update_progress(
         &mut self,
         vnode: VirtualNode,
-        progress: BackfillProgressPerVnode,
-    ) -> Option<BackfillProgressPerVnode> {
-        self.inner.insert(vnode, progress)
+        new_pos: OwnedRow,
+    ) -> StreamExecutorResult<()> {
+        let state = self.get_current_state(&vnode);
+        let new_state = BackfillProgressPerVnode::InProgress(new_pos);
+        match state {
+            BackfillProgressPerVnode::NotStarted => *state = new_state,
+            BackfillProgressPerVnode::InProgress(_current_pos) => *state = new_state,
+            BackfillProgressPerVnode::Completed { .. } => unreachable!(),
+        }
+        Ok(())
     }
 
-    fn iter_backfill_progress(
-        &self,
-    ) -> impl Iterator<Item = (&VirtualNode, &BackfillProgressPerVnode)> {
-        self.inner.iter()
+    pub(crate) fn finish_progress(&mut self, vnode: VirtualNode, pos_len: usize) {
+        let finished_placeholder_position = construct_initial_finished_state(pos_len);
+        let current_state = self.get_current_state(&vnode);
+        let new_pos = match current_state {
+            BackfillProgressPerVnode::NotStarted => finished_placeholder_position,
+            BackfillProgressPerVnode::InProgress(current_pos) => current_pos.clone(),
+            BackfillProgressPerVnode::Completed { .. } => {
+                return;
+            }
+        };
+        *current_state = BackfillProgressPerVnode::Completed(new_pos);
+    }
+
+    /// If can commit, return state to be committed.
+    fn get_commit_state(&self, vnode: &VirtualNode) -> Option<(Option<Vec<Datum>>, Vec<Datum>)> {
+        if !self.can_commit(vnode) {
+            return None;
+        }
```

Comment:
Moving this code outside the `get_commit_state` would make it more straightforward.